src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -11,177 +11,132 @@ import breeze.numerics._
-      val resBc = sc.broadcast(res)
-      val CinvXtVBc = sc.broadcast(Cinv * VX.t)
-
-      def preprocessGenotypes(gs: Vector[Double], w: Double): SkatTuple = {
-        val wx = gs * math.sqrt(w)
-        val sj = resBc.value dot wx
-        val CinvXtVwx = CinvXtVBc.value * wx
-        SkatTuple(sj * sj, wx :* sqrtVBc.value , CinvXtVwx)
-      }
-
-      val skatRDD = keyedRdd
-        .map { case (key, vs) =>
-          val vArray = vs.toArray.map { case (gs, w) => preprocessGenotypes(gs, w) }
-          val skatStat = resultOp(vArray, 2.0)
-          Row(key, skatStat.q, skatStat.p, skatStat.fault)
-        }
-
-      val skatSignature = SkatStat.schema(keyType)
-
-      new KeyTable(vds.hc, skatRDD, skatSignature, Array("key"))
-    }
-
-    val completeSamplesBc = filteredVds.sparkContext.broadcast((0 until n).toArray)
 
-    val getGenotypesFunction = (gs: Iterable[Genotype], n: Int) =>
-      if (!useDosages) {
-        RegressionUtils.hardCalls(gs, n)
-      } else {
-        RegressionUtils.dosages(gs, completeSamplesBc.value)
-      }
-
-    val (keyedRdd, keysType) = keyedRDDSkat(filteredVds, variantKeys, singleKey, weightExpr, getGenotypesFunction)
+    val filteredVds = vds.filterSamplesMask(sampleMask)
+    
+    val (keyGsWeightRdd, keyType) = toKeyGsWeightRdd(filteredVds, variantKeys, singleKey, weightExpr, useDosages)
 
-    if (!logistic) computeSkatLinear(keyedRdd, keysType, y, cov, if (!useLargeN) runSkatPerKey else runSkatPerKeyLargeN)
-    else computeSkatLogistic(keyedRdd, keysType, y, cov, if (!useLargeN) runSkatPerKey else runSkatPerKeyLargeN)
+    val skatRdd: RDD[Row] = 
+      if (logistic)
+        logisticSkat(keyGsWeightRdd, y, cov, maxSize, accuracy, iterations, forceLargeN)
+      else
+        linearSkat(keyGsWeightRdd, y, cov, maxSize, accuracy, iterations, forceLargeN)
+    
+    val skatSignature = TStruct(
+      ("key", keyType),
+      ("size", TInt32),
+      ("qstat", TFloat64),
+      ("pval", TFloat64),
+      ("fault", TInt32))
+
+    new KeyTable(vds.hc, skatRdd, skatSignature, Array("key"))
   }
-
-  def keyedRDDSkat(vds: VariantDataset,
+  
+  def toKeyGsWeightRdd(vds: VariantDataset,
     variantKeys: String,
     singleKey: Boolean,
     weightExpr: Option[String],
-    getGenotypes: (Iterable[Genotype], Int) => Vector[Double]):
+    useDosages: Boolean):
   (RDD[(Annotation, Iterable[(Vector[Double], Double)])], Type) = {
+    // ((key, [(gs_v, weight_v)]), keyType)
     
     val n = vds.nSamples
     
     val vdsWithWeight =
       if (weightExpr.isEmpty)
         vds.annotateVariantsExpr("va.__AF = gs.callStats(g => v).AF")
-          .annotateVariantsExpr("va.__weight = let af = if (va.__AF[0] <= va.__AF[1]) va.__AF[0] else va.__AF[1] in dbeta(af, 1.0, 25.0)**2")
+          .annotateVariantsExpr("va.__weight = let af = " +
+            "if (va.__AF[0] <= va.__AF[1]) va.__AF[0] else va.__AF[1] in dbeta(af, 1.0, 25.0)**2")
       else
         vds
-
+    
```

Comment:
Will remove this instance so it's not in the diff, but in general I don't think we're consistent on whether there is indenting whitespace up to text alignment within a class or def. Do you agree?

Reply:
I'm not sure what you mean. General practice is to never include trailing white-space since it doesn't affect formatting but adds bytes to the file (this doesn't really matter these days).